### DISCUSSION TITLE
Display Enum drop-down when I use Depends

### DISCUSSION THREAD

**Discussion by vicente23** [fastapi/fastapi#9266] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from enum import Enum

from fastapi import  Depends, FastAPI, HTTPException

app = FastAPI()

class Names(str, Enum):
    Bob = "Bob"
    Jack = "Jack"

async def name_parameter(name: str) -> Names:
    try:
        return Names(name)
    except ValueError:
        raise HTTPException(detail="wrong name", status_code=422)

@app.get("/")
async def root(name_valid_representation: Names,
               name_wrong_representation: Names = Depends(name_parameter)
):
    return {"message": f"Hello"}
```

### Description

For my application, I use several  `Enum`. 

I need to customize the error messages automatically returned by pydantic.

For this, I use a parameter of type `Enum` and the `Depends` method which allows me to customize the error message using the `HTTPException` method. 

This works, but I noticed that when using Depends, the dropdown list with the allowed values no longer appears. 

Is there a way to display this dropdown list while customizing the error message? 

![image](https://user-images.githubusercontent.com/36680163/225895464-db5bb4bb-ca94-4800-8c26-fb34fe1f972a.png)

I also explored `Literal` typing as well, same result.

Thanks in advance for your answers 

### Operating System

Linux

### FastAPI Version

0.94.0

### Python Version

3.10.6

**Comment by jgould22** [comment, marked as answer]
The reason is because your dependency isnt taking an Enum any more it is taking a string so the openapi generator reflects that.

A better approach might be to write a custom exception handler to handle the exception from pydantic to return what you would prefer.

You can read some in the docs here https://fastapi.tiangolo.com/tutorial/handling-errors/#override-request-validation-exceptions

Then you wont need the dependency and your endpoint can simply list the enum type like you do with name_valid_representation.

**Reply by vicente23** [reply]
Thanks for the explanation @jgould22.

I try to implement the example of the documentation, it works. I can easily customize the error message. To extract the first parameter that is invalid, I use the `exc.errors()[0]['loc'][1]`. Do you know a more efficient method ? 

Here is the method that allows me to identify the invalid parameter:

```python
@app.exception_handler(RequestValidationError)
async def validation_exception_handler(_request, exc):
    return PlainTextResponse(json.dumps({"message": f"The parameter {exc.errors()[0]['loc'][1]} is invalid"}), status_code=400)
```

Another way is to override the openapi documentation when starting the API by specifying the `enum` property for my parameter in the `openapi.Json` file. I get the expected result in the Swagger documentation. However, this is not ideal because I want to use `pydantic` and it is more difficult to maintain in my opinion. 

Thanks!

**Comment by smarie** [comment]
I have the same issue than you, with all kind of usages of Enums. The Swagger UI autodoc always sets the dropdown menu as "disabled".

For example this :

```python
from fastapi import FastAPI
from enum import Enum

app = FastAPI()

class Country(str, Enum):
    eu = "eu"
    us = "us"

@app.get("/")
def get_something(country: Country = Country.eu):
    return {"country": country.value}
```

This is the screenshot. You cannot see it but hovering over the dropdown leads my mouse icon to turn into a "forbidden" icon.

![image](https://user-images.githubusercontent.com/3236794/226112163-69ba2167-6e37-4ad2-a550-ff105068993e.png)

The same happens for the example in the doc : https://fastapi.tiangolo.com/tutorial/path-params/?h=enum#working-with-python-enumerations

![image](https://user-images.githubusercontent.com/3236794/226112254-1b37bf92-6f1f-42a0-a866-0e3f0f3ecd45.png)

I tried to do the same with an explicit `Form(enum=...)` but same result.
 
My versions:
 - FastAPI: 0.94.1
 - Uvicorn: 0.21.1
 - Python: 3.9.6

**Reply by jgould22** [reply]
> I have the same issue than you, with all kind of usages of Enums. The Swagger UI autodoc always sets the dropdown menu as "disabled".

It is not really the same issue, that is the behavior of swagger doc.

Just click "try it out" button the right side of that panel and it will unlock the field.

**Reply by smarie** [reply]
Indeed ! Oh, sorry for annoying you with such an obvious "problem". Thanks for the quick support !
